### DISCUSSION TITLE
Rewrite error message for `/keys` endpoint when mellisearch is run without a master key

### DISCUSSION THREAD

**Discussion by akhildevelops** [meilisearch/product#552] [Feedback & Feature Proposal]
GET `/keys` request fails on `missing_authorization_header` when Mellisearch is run without masterkey. This error makes sense only if Mellisearch is run with a master key.

**Original Error message**
```
{
    "message": "The Authorization header is missing. It must use the bearer authorization method.",
    "code": "missing_authorization_header",
    "type": "auth",
    "link": "https://docs.meilisearch.com/errors#missing_authorization_header"
}
```
It would easier to understand if the message can be shown as below
**Expected Error message**

The error message can be:
```
{
    "message": "Mellisearch is running without Master Key. Set the Master Key by following: https://docs.meilisearch.com/learn/security/master_api_keys.html#protecting-a-meilisearch-instance",
    "code": "no_master_key",
    "type": "auth",
    "link": "https://docs.meilisearch.com/errors#no_master_key"
}
```

**Comment by akhildevelops** [comment]
Even the docs suggest to set master key for interacting with Keys. https://docs.meilisearch.com/reference/api/keys.html#keys

**Comment by curquiza** [comment]
Hello @Enforcer007 

Thanks for the suggestion

I think indicating the master key is not set could be really dangerous, but I might be wrong, WDYT @gmourier and @Kerollmops?

**Reply by Kerollmops** [reply]
We talked about that with @ManyTheFish, and thought that adding an error message that clearly explains the issue here isn't dangerous as an attacker will always try the dangerous routes. An instance without a master key is already running with zero security, so there is no way to make it more insecure.

However, one advantage of having a specific error message is to help the user understand what's wrong. I would see a new error code, `missing_master_key`, and an error message that looks like the one proposed.

Thank you @Enforcer007 for the proposal 👍

**Reply by gmourier** [reply]
I agree with @Kerollmops. Do we consider it to be an `auth` error here? If we consider it, `401` seems more suited than `403` imo.

Thanks, @Enforcer007.

**Reply by Kerollmops** [reply]
Indeed, I would say that [a _401_ response code](http://web.archive.org/web/20210802131115/https://httpstatuses.com/401) is more suited than [a _403_ response code](http://web.archive.org/web/20210802131115/https://httpstatuses.com/403) when the master key is missing.

**Comment by akhildevelops** [comment]
Hi @curquiza ,

I think it shouldn't be worrisome as :

- mellisearch will be mostly run in non production environments without setting master key.
- setting master key is enforced when running in a production environment therefore the _Expected Error_ message will never occur.

Expected error message:
```
The error message can be:

{
    "message": "Mellisearch is running without Master Key. Set the Master Key by following: https://docs.meilisearch.com/learn/security/master_api_keys.html#protecting-a-meilisearch-instance",
    "code": "no_master_key",
    "type": "auth",
    "link": "https://docs.meilisearch.com/errors#no_master_key"
}

```

**Comment by curquiza** [comment]
Hello @Enforcer007 @gmourier and @Kerollmops

I moved this discussion to the product side for consistency (definitely a product discussion)
I created the corresponding spec to it: https://github.com/meilisearch/specifications/pull/179

Everyone is welcome to review 😄 
Once it will be approved, I will open an issue on Meilisearch side!

**Comment by gmourier** [comment]
Hey everyone 👋

I'm locking that discussion since it has been implemented with `v0.30`

Please open a new discussion if it's needed.

Thank you 🙇‍♂️
